dot/parachain/candidate-validation/candidate_validation.go

```diff
@@ -1,25 +1,101 @@
-// Copyright 2023 ChainSafe Systems (ON)
+// Copyright 2024 ChainSafe Systems (ON)
 // SPDX-License-Identifier: LGPL-3.0-only
 
-package parachain
+package candidatevalidation
 
 import (
 	"bytes"
+	"context"
 	"errors"
 	"fmt"
+	"sync"
 
 	parachainruntime "github.com/ChainSafe/gossamer/dot/parachain/runtime"
 	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/internal/log"
 	"github.com/ChainSafe/gossamer/lib/common"
-
 	"github.com/ChainSafe/gossamer/pkg/scale"
 )
 
+var logger = log.NewFromGlobal(log.AddContext("pkg", "parachain-candidate-validation"))
+
 var (
 	ErrValidationCodeMismatch   = errors.New("validation code hash does not match")
 	ErrValidationInputOverLimit = errors.New("validation input is over the limit")
 )
 
+type CandidateValidation struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
```

Comment:
I'd like to refactor this to remove context and cancel since we agree that context shouldn't be used for cancel.  I've looked at https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation however the solution outlined there does not handle separate go routines.  We've talked about service archecture, do we have an implemented example I can model this from? Any suggestions for removing context from subsystem struct, and handling life scope?  @timwu20 @kishansagathiya

Reply:
One alternative could be using channels instead of contexts. The struct will hold a `stopCh chan struct{}` the purpose of this channel is to be closed once the `Stop` method is called.

```go
type Str struct {
  wg sync.WaitGroup
  stopCh chan struct{}
}

func (s *Str) Run() {
  wg.Add(1)
  go subprocess(&wg)
}

func (s *Str) Stop() {
   close(s.stopCh)
   s.wg.Wait()
}


func subprocess(wg *sync.WaitGroup) {
  defer wg.Done()

  for {
      select {
      case <-someOtherChan:
         ...
      case <-stopCh:
         return
      }
  }
}
```

Reply:
Yes, I've implemented using this model.